### PR TITLE
chore (refs T34979): add DpEditor to Storybook.

### PR DESCRIPTION
**Pull request by ahmad-demos**
https://yaits.demos-deutschland.de/T34979
https://github.com/demos-europe/demosplan-core/pull/2132

- add DpEditor to Storybook.
- add missing props, when editor used dpUploadModal.

### CODE REVIEW

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 1]
```diff
@@ -16,4 +16,7 @@ A rich text editor component based on tiptap.
 - the editor-content component needs a prop with editor instance to work correctly
 
 To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+
+To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically.
+
+The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
```
Comment (salisdemos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:22]: this is dplan specific. So it should not be mentioned here.
But maybe we can add a configuration where the slots are used. Something like the imageUpload modal ?

**Review by salisdemos — CHANGES_REQUESTED** [review]

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 2]
```diff
@@ -16,4 +16,7 @@ A rich text editor component based on tiptap.
 - the editor-content component needs a prop with editor instance to work correctly
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:16]: These content above is tailored to the specific implementation within demosplan. I'd suggest to delete all of it.

**Review by spiess-demos — CHANGES_REQUESTED** [review]

File: src/components/core/DpEditor/DpEditor.stories.mdx (hunk 1 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/df50c169fd1b50ab220d9f778e973311f2f03123

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 3]
```diff
@@ -16,4 +16,7 @@ A rich text editor component based on tiptap.
...
 
 To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+
+To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically.
```
[the remark this answers is not in the document]
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/df50c169fd1b50ab220d9f778e973311f2f03123

File: src/components/core/DpEditor/DpEditor.stories.mdx (hunk 2 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/df50c169fd1b50ab220d9f778e973311f2f03123

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 4]
```diff
@@ -13,7 +13,9 @@ A rich text editor component based on tiptap.
 - defines a menubar with a number of buttons and an editor
 - use this component without the inline-editing-wrapper TiptapEditText.vue if you want to add a text editor to a form element (as in new statement view)
 - use this component with the inline-editing-wrapper TiptapEditText.vue if you want to save the text directly via inline-editing (as in assessment table)
-- the editor-content component needs a prop with editor instance to work correctly
 
 To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+
+To use image upload we need to add imageButton Prop.
```
Comment (salisdemos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:19]: If You write that, You should metion all optional props.
And therefor we should better introduce a table or so

Comment (salisdemos) [reply src/components/core/DpEditor/DpEditor.stories.mdx:19]: and we still don't have a example how it works with the usage of slots.

**Review by salisdemos — CHANGES_REQUESTED** [review]

File: src/components/core/DpEditor/DpEditor.stories.mdx (hunk 4 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/428ca8f7b5f5e5120e633ef18f2d41ba67b8c8ac
added optional props in Custom usage.

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 5]
```diff
@@ -11,9 +11,61 @@ import DpEditor from './DpEditor'
...
+    <template v-slot:button>
+        <button
+            class="{{ 'menubar__button'|prefixClass }}"
+            type="button"
+            v-tooltip="'{{ 'boilerplate.insert'|trans }})'"
+            @click.stop="$refs.boilerPlateModal.toggleModal()">
+            <i class="{{ 'fa fa-puzzle-piece'|prefixClass }}"></i>
+        </button>
+    </template>
+</dp-editor>
```
Comment (salisdemos) [inline src/components/core/DpEditor/DpEditor.stories.mdx]: shouldn't this be a separat tsx file?

Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: I think not. which one exactly?
I deleted v-slot:button !

Comment (salisdemos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: I was thinking of a tsx-file with an example of a dpEditor-Configuration with a working image upload modal

Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/e6411cdd2e790f2a79557ba2a01dbc0088eb0d9c

Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/734df709a61a8d11da4b0fdbaa9df31b48906b0b

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 6]
```diff
@@ -11,9 +11,42 @@ import DpEditor from './DpEditor'
...
-- use this component without the inline-editing-wrapper TiptapEditText.vue if you want to add a text editor to a form element (as in new statement view)
-- use this component with the inline-editing-wrapper TiptapEditText.vue if you want to save the text directly via inline-editing (as in assessment table)
-- the editor-content component needs a prop with editor instance to work correctly
 
 To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+
+To use image upload we need to add imageButton optional Prop.
+
+To use another component within DpEditor, we can add Slot named modal.
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:19]: This implies that DpEditor can be easily extended using slots. This is not the case - `v-slot:modal` is still tightly coupled to the inner workings of the boilerplate feature. I'd suggest no not mention the slot functionality at all, at least before we have a fully working way of extending DpEditor.

**Review by spiess-demos — CHANGES_REQUESTED** [review]

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 7]
```diff
@@ -11,9 +11,42 @@ import DpEditor from './DpEditor'
...
 
 - defines a menubar with a number of buttons and an editor
-- use this component without the inline-editing-wrapper TiptapEditText.vue if you want to add a text editor to a form element (as in new statement view)
-- use this component with the inline-editing-wrapper TiptapEditText.vue if you want to save the text directly via inline-editing (as in assessment table)
-- the editor-content component needs a prop with editor instance to work correctly
 
 To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+
+To use image upload we need to add imageButton optional Prop.
```
[the remark this answers is not in the document]
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/e6899199500bcb6290964231adf96b76dfe2a5b4
Sorry, wrong commit naming

File: src/components/core/DpEditor/DpEditor.stories.mdx (hunk 6 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/e6899199500bcb6290964231adf96b76dfe2a5b4
Sorry, wrong commit naming

Comment (spiess-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: i meant to not mention it at all.

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 8]
```diff
@@ -8,12 +8,46 @@ import DpEditor from './DpEditor'
...
+        getFileByHash: (hash) => RoutingFunction(hash)
+    }"
+    :value="value">
+    <template v-slot:modal="modalProps">
+        <dp-modal
+          id="openModal"
+          ref="openModal">
+          Add an component
+        </dp-modal>
+    </template>
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:51]: Well, if we really want to document this case here, we'd have to show it all, including the usage of `handleInsertText` exposed by DpEditor, and the button which triggers the modal.

```suggestion
    <template v-slot:modal="modalProps">
        <dp-modal
            id="editorModal"
            ref="editorModal"
            @insert="text => modalProps.handleInsertText(textFromParentComponent)">
            Add an component
        </dp-modal>
    </template>
    <template v-slot:button>
        <dp-button
              :text="Add Text"
              variant="subtle"
              @click.stop="openEditorModal" />
    </template>
```

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 9]
```diff
@@ -8,12 +8,46 @@ import DpEditor from './DpEditor'
...
-- use this component without the inline-editing-wrapper TiptapEditText.vue if you want to add a text editor to a form element (as in new statement view)
-- use this component with the inline-editing-wrapper TiptapEditText.vue if you want to save the text directly via inline-editing (as in assessment table)
-- the editor-content component needs a prop with editor instance to work correctly
 
-To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+To use image upload we need to add imageButton optional Prop, and when using `imageButton` Prop, you have to set also
+`routes.getFileByHash` Prop.
+
+To use another component within DpEditor, we can add Slot.
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:17]:
```suggestion
The slot `button` can be used to add additional buttons to the toolbar. The slot `modal` can be used to add a modal component to DpEditor. The function `handleInsertText` is exposed to the `modal` slot to enable the modal component to add text to the editor instance.
```

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 10]
```diff
@@ -8,12 +8,46 @@ import DpEditor from './DpEditor'
...
 
-- defines a menubar with a number of buttons and an editor
-- use this component without the inline-editing-wrapper TiptapEditText.vue if you want to add a text editor to a form element (as in new statement view)
-- use this component with the inline-editing-wrapper TiptapEditText.vue if you want to save the text directly via inline-editing (as in assessment table)
-- the editor-content component needs a prop with editor instance to work correctly
 
-To properly set the content we have to update this.currentValue and use editor.setContent() - both actions are needed!
-To use boilerplates we need to mount tiptap editor, where boilerplate modal is imported dynamically. The boilerplates are stored in boilerplates vuex store, which is also registered dynamically if boilerplate prop is specified.
+To use image upload we need to add imageButton optional Prop, and when using `imageButton` Prop, you have to set also
+`routes.getFileByHash` Prop.
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.mdx:15]:
```suggestion
Uploading images can be enabled via the imageButton prop. When using `imageButton`, the `routes.getFileByHash` prop must be set to an existing route which returns a file by its hash (see second example).
.
```

**Review by spiess-demos — CHANGES_REQUESTED** [review]

File: src/components/core/DpEditor/DpEditor.stories.mdx (hunk 6 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/2fc10b706e8b427d0a2e8ff420e209799380f9b7
see hier, I added also missing props!

File: src/components/core/DpEditor/DpEditor.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
+import type { Meta, StoryObj } from '@storybook/vue'
+import DpEditor from './DpEditor.vue'
+import DpModal from '../../DpModal/DpModal.vue'
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.tsx:3]:
```suggestion
import DpModal from '~/components/DpModal'
```

File: src/components/core/DpEditor/DpEditor.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        },
+        template: `<dp-editor v-bind="args"></dp-editor>`,
+    })
+}
+
+interface IDpEditor {
+    tusEndpoint: string
+    value: string
+    toolbarItems: object
+    routes: object
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.tsx:23]: Nitpick: alpnabetical sorting

File: src/components/core/DpEditor/DpEditor.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            insertAndDelete: true,
+            table: true,
+            headings: [2,3,4],
+            imageButton: true,
+            linkButton: true,
+            strikethrough: true
+        },
+        routes: {
+            getFileByHash: () => {}
+        }
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.tsx:52]: Nitpick: alphabetical sorting

File: src/components/core/DpEditor/DpEditor.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            return { args }
+        },
+        template: `<dp-editor v-bind="args">
+          <template v-slot:modal="modalProps">
+            <dp-modal
+                id="openModal"
+                ref="openModal">
+              Add an component
+            </dp-modal>
+          </template>
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.stories.tsx:69]: This is not a working example, as it does nothing. Either we implement real logic here (with a button triggering a modal, from which text can be inserted), or we only use the simple example here.

File: src/components/core/DpEditor/DpEditor.vue [hunk 15]
```diff
@@ -9,9 +9,11 @@
       ref="linkModal"
       @insert="insertUrl" />
     <dp-upload-modal
-      v-if="toolbar.imageButton"
+      v-if="toolbar.imageButton && tusEndpoint"
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.vue:12]: While it is ok to render the modal only when all needed props are given, this also leads to a silent error if a developer is not aware of the fact she has to set tusEndpoint when using imageButton. I'd suggest to add the following in the `mounted` hook:

```
if (toolbar.imageButton ^ tusEndpoint) {
  console.warn(`DpEditor is called olny one of toolbar.imageButton or tusEndpoint set. Both must be used.`)
}
```

Note that the [^ operator (bitwise XOR) ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_XOR) is just a condensed expression of this:

```
(toolbar.imageButton && !tusEndpoint || !toolbar.imageButton && tusEndpoint)
```

File: src/components/core/DpEditor/DpEditor.vue [hunk 16]
```diff
@@ -373,6 +375,12 @@ export default {
   mixins: [prefixClassMixin],
 
   props: {
+    basicAuth: {
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpEditor.vue:378]: Can you add a short docblock here, too?

File: src/components/core/DpEditor/DpUploadModal.vue [hunk 17]
```diff
@@ -66,9 +68,23 @@ export default {
   },
 
   props: {
+    basicAuth: {
```
Comment (spiess-demos) [inline src/components/core/DpEditor/DpUploadModal.vue:71]: Can you add a short docblock here, too?

**Review by spiess-demos — CHANGES_REQUESTED** [review]

File: src/components/core/DpEditor/DpEditor.stories.mdx [hunk 18]
```diff
@@ -8,12 +8,57 @@ import DpEditor from './DpEditor'
...
+
+The slot `modal` can be used to add a modal component to DpEditor.
+
+The function `handleInsertText` is exposed to the `modal` slot to enable the modal component to add text to the editor instance.
+
+## Default usage
+
+```html
+<dp-editor
+    :value="value"
```
[the remark this answers is not in the document]
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.mdx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/55c3ae63b443d85b1ea29de3ed94bf27d8193438

File: src/components/core/DpEditor/DpEditor.stories.tsx (hunk 12 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.stories.tsx]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/55c3ae63b443d85b1ea29de3ed94bf27d8193438

File: src/components/core/DpEditor/DpEditor.vue (hunk 16 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.vue:381]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/55c3ae63b443d85b1ea29de3ed94bf27d8193438

File: src/components/core/DpEditor/DpUploadModal.vue (hunk 17 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpUploadModal.vue:74]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/55c3ae63b443d85b1ea29de3ed94bf27d8193438

File: src/components/core/DpEditor/DpEditor.vue (hunk 15 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.vue:12]: https://github.com/demos-europe/demosplan-ui/pull/567/commits/12995fdabbd6446d4924f8e9de9e4bb1c5638907

File: src/components/core/DpEditor/DpEditor.vue (hunk 16 above)
Comment (spiess-demos) [reply src/components/core/DpEditor/DpEditor.vue:381]: I'd suggest to change the description to "The Tus endpoint requires basicAuth to be added to the file header.".

File: src/components/core/DpEditor/DpUploadModal.vue (hunk 17 above)
Comment (spiess-demos) [reply src/components/core/DpEditor/DpUploadModal.vue:74]: See other comment.

File: src/components/core/DpEditor/DpEditor.vue (hunk 16 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpEditor.vue:381]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/c3b38316ab1710741b6f7d284e3a54b75865b3fa

File: src/components/core/DpEditor/DpUploadModal.vue (hunk 17 above)
Comment (ahmad-demos) [reply src/components/core/DpEditor/DpUploadModal.vue:74]: Done in https://github.com/demos-europe/demosplan-ui/pull/567/commits/c3b38316ab1710741b6f7d284e3a54b75865b3fa

**Review by spiess-demos — APPROVED** [review]
my concerns are all addressed. however as i contributed here, too, @salisdemos should approve the PR, too.

**Review by salisdemos — APPROVED** [review]

**Pull request merged**